### PR TITLE
Repository Structure Implementation

### PR DESCRIPTION
**Pull request by adomas-vensas**
Adding initial project structure and classes for CRUD operations:
- Added a `Controllers` folder to `Recipes.API`
- Added a `Recipes.Application` project
- Added `Services`, `Interfaces`, `DTOs` folders to `Recipes.Application`
- Injected `IRecipesService` and `RecipesService` into the API
- Added DTOs for Recipe creation and update

This pull request is for #9 issue.

### CODE REVIEW

File: Recipes/Recipes.Application/Services/RecipesService.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+using Recipes.Public;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Recipes.Application.Services;
+
+public class RecipesService : IRecipesService
```
Comment (6ar8nas) [inline Recipes/Recipes.Application/Services/RecipesService.cs:11]: I wonder if we need this here. We have created our own client-sided RecipeService, that consumes the API in FrontEnd.Web project through HttpClient requests.

File: Recipes/Recipes.Application/DTOs/Recipes/RecipeUpdateDTO.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Recipes.Application.DTOs.Recipes;
+
+public class RecipeUpdateDTO
+{
```
Comment (6ar8nas) [inline Recipes/Recipes.Application/DTOs/Recipes/RecipeUpdateDTO.cs:10]: I suppose the DTOs could also go to the Recipes.Public project. It is after all the public facing object representation format.

File: Recipes/Recipes.Application/DTOs/Recipes/RecipeCreationDTO.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
```
Comment (6ar8nas) [inline Recipes/Recipes.Application/DTOs/Recipes/RecipeCreationDTO.cs:1]: Not a big issue, but I think many of these imports are unused and not only in this particular class

File: Recipes/Recipes.API/Controllers/RecipesController.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using Recipes.Application.Services;
+
+namespace Recipes.API.Controllers;
+
+[ApiController]
+[Route("recipes")]
+public class RecipesController : ControllerBase
+{
+    private IRecipesService _recipesService;
```
Comment (6ar8nas) [inline Recipes/Recipes.API/Controllers/RecipesController.cs:10]: Probably this should be some RecipesRepository instead. Haven't looked through the whole source too thoroughly, but the rough idea would probably be using a [repository pattern](https://medium.com/@codebob75/repository-pattern-c-ultimate-guide-entity-framework-core-clean-architecture-dtos-dependency-6a8d8b444dcb)

**Review by 6ar8nas — CHANGES_REQUESTED** [review]

File: Recipes/Recipes.Application/DTOs/Recipes/RecipeUpdateDTO.cs (hunk 2 above)
Comment (adomas-vensas) [reply Recipes/Recipes.Application/DTOs/Recipes/RecipeUpdateDTO.cs]: Yup, seems logical

File: Recipes/Recipes.API/Controllers/RecipesController.cs (hunk 4 above)
Comment (adomas-vensas) [reply Recipes/Recipes.API/Controllers/RecipesController.cs:10]: So, as I understand, we don't really need `services` in the back-end as you are already handling them? We should only have `repositories` instead?

Comment (6ar8nas) [reply Recipes/Recipes.API/Controllers/RecipesController.cs:10]: Yeah, I'd advocate for using just the repository as an intermediary layer between the database and the controller. Noticed these docs which seem to be more official and trustworthy, which had this nice illustration (our pathway would be the second):

Comment (adomas-vensas) [reply Recipes/Recipes.API/Controllers/RecipesController.cs:10]: All right, I will implement this instead 👌

**Comment by adomas-vensas** [comment]
Updated the project structure to adhere to the Repository pattern as pointed out by @6ar8nas

File: Recipes/Recipes.DataAccess/Interfaces/IGenericRepository.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Recipes.DataAccess.Interfaces;
+
+public interface IGenericRepository<T> where T : class
+{
+    public Task<T> CreateAsync(T entity);
```
Comment (6ar8nas) [inline Recipes/Recipes.DataAccess/Interfaces/IGenericRepository.cs:5]:
```suggestion
    public Task<T> InsertAsync(T entity);
```

Nitpick, I think it's just more common, perhaps due to SQL syntax.

File: Recipes/Recipes.DataAccess/Interfaces/IGenericRepository.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+    public Task<T> GetByIdAsync(int id);
+    public Task<T> GetAllAsync();
+    public Task<T> DeleteByIdAsync(int id);
```
Comment (6ar8nas) [inline Recipes/Recipes.DataAccess/Interfaces/IGenericRepository.cs:8]: You probably wouldn't return the no longer existant object.

```suggestion
    public Task DeleteByIdAsync(int id);
```

File: Recipes/Recipes.DataAccess/Interfaces/IGenericRepository.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Recipes.DataAccess.Interfaces;
+
+public interface IGenericRepository<T> where T : class
+{
+    public Task<T> CreateAsync(T entity);
+    public Task<T> GetByIdAsync(int id);
+    public Task<T> GetAllAsync();
```
Comment (6ar8nas) [inline Recipes/Recipes.DataAccess/Interfaces/IGenericRepository.cs:7]:
```suggestion
    public Task<IEnumerable<T>> GetAllAsync();
```

File: Recipes/Recipes.DataAccess/Interfaces/IGenericRepository.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Recipes.DataAccess.Interfaces;
+
+public interface IGenericRepository<T> where T : class
```
Comment (6ar8nas) [inline Recipes/Recipes.DataAccess/Interfaces/IGenericRepository.cs:3]: `UpdateAsync` is missing.

File: Recipes/Recipes.DataAccess/Repositories/GenericRepository.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    private readonly DbContext _dbContext;
+    private readonly DbSet<T> _dbSet;
+
+    public GenericRepository(DbContext dbContext, DbSet<T> dbSet)
+    {
+        _dbContext = dbContext;
+        _dbSet = dbSet;
+    }
+
+    public async Task<T> CreateAsync(T entity)
```
Comment (6ar8nas) [inline Recipes/Recipes.DataAccess/Repositories/GenericRepository.cs:17]: Regarding the warnings in GitHub - I think you can just remove the async keywords for now from here until they are implemented.

File: Recipes/Recipes.API/Controllers/RecipesController.cs [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+[Route("recipes")]
+public class RecipesController : ControllerBase
+{
+    private readonly IGenericRepository<Recipe> _recipesRepository;
+
+    public RecipesController(IGenericRepository<Recipe> recipesRepository)
+    {
+        _recipesRepository = recipesRepository;
+    }
+
```
Comment (6ar8nas) [inline Recipes/Recipes.API/Controllers/RecipesController.cs:17]: Could we move the `/recipes` endpoint to the controller from Program.cs with this PR already? We would make sure everything works as expected.

**Review by 6ar8nas — CHANGES_REQUESTED** [review]

File: Recipes/Recipes.Application/Services/RecipesService.cs (hunk 1 above)
Comment (6ar8nas) [reply Recipes/Recipes.Application/Services/RecipesService.cs:6]: I noticed you have resolved this conversation, but we still have the Recipes.Application project. Do we want to keep it?

File: Recipes/Recipes.API/Controllers/RecipesController.cs (hunk 10 above)
Comment (adomas-vensas) [reply Recipes/Recipes.API/Controllers/RecipesController.cs:17]: Sure, going to do that right now

Comment (adomas-vensas) [reply Recipes/Recipes.API/Controllers/RecipesController.cs:17]: I implemented the generic repository functions and added a single endpoint to get all recipes in the controller. However, I seem to be getting an exception when trying to retrieve the recipes. The exception happens even before it enters the endpoint:

paveikslas

I re-ran the `docker-compose` file and tried to update the database. I got this error:

paveikslas

File: Recipes/Recipes.Application/Services/RecipesService.cs (hunk 1 above)
Comment (adomas-vensas) [reply Recipes/Recipes.Application/Services/RecipesService.cs]: Sorry, I thought I removed the `Application` layer previously. We definitely don't need it anymore.

File: Recipes/Recipes.API/Controllers/RecipesController.cs (hunk 10 above)
Comment (6ar8nas) [reply Recipes/Recipes.API/Controllers/RecipesController.cs:29]: DbSet has previously been injected to the repository, which couldn't be resolved. I pushed a commit that fixes it.

File: Recipes/Recipes.API/Controllers/RecipesController.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        try
+        {
+            recipes = await recipesRepository.GetAllAsync();
+        }
+        catch (Exception)
+        {
+            return StatusCode(StatusCodes.Status500InternalServerError);
+        }
+        
+        // Return type currently doesn't match actual type. RecipeEntity should be probably mapped to some RecipeResponse DTO type.
```
Comment (6ar8nas) [inline Recipes/Recipes.API/Controllers/RecipesController.cs:28]: We currently have `RecipeEntity` database entity class and we probably consider just `Recipe` to be our DTO? Currently I didn't notice any differences between those two classes, but I think we should return the database entity mapped to a public class object.

File: Recipes/Recipes.API/Controllers/RecipesController.cs [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        try
+        {
+            recipes = await recipesRepository.GetAllAsync();
+        }
+        catch (Exception)
+        {
+            return StatusCode(StatusCodes.Status500InternalServerError);
+        }
+        
+        // Return type currently doesn't match actual type. RecipeEntity should be probably mapped to Recipe.
```
Comment (6ar8nas) [inline Recipes/Recipes.API/Controllers/RecipesController.cs:27]: We currently have `RecipeEntity` database entity class and we probably consider just `Recipe` to be our DTO. Currently I didn't notice any differences between those two classes, but I think we should return the database entity mapped to a public class object.

Comment (adomas-vensas) [reply Recipes/Recipes.API/Controllers/RecipesController.cs:27]: Can i map it to a regular `Recipe`?

Comment (adomas-vensas) [reply Recipes/Recipes.API/Controllers/RecipesController.cs:27]: Oh, nevermind. All right. I somehow missed the comment part

Comment (adomas-vensas) [reply Recipes/Recipes.API/Controllers/RecipesController.cs]: All right, I mapped it. Seems to work

**Review by 6ar8nas — APPROVED** [review]
Looks good to me, tested in the frontend - seems to work fine. I guess a second approval in case I missed anything would be nice, but worst case we can solve those issues with a follow-up PR.

**Comment by adomas-vensas** [comment]
Okay, then I'll merge and move onto the rest of the CRUD operations.

**Pull request merged**